builder/bundle_control_test.go

```diff
@@ -0,0 +1,432 @@
+package builder
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"reflect"
+	"testing"
+
+	"github.com/clearlinux/mixer-tools/config"
+)
+
+func TestGetUpstreamBundlesVerDir(t *testing.T) {
+	testCases := []struct {
+		ver string
+		exp string
+	}{
+		{"0", "clr-bundles-0"},
+		{"000001", "clr-bundles-000001"},
+		{"non-numeric", "clr-bundles-non-numeric"},
+		{"", "clr-bundles-"},
+		{"25660", "clr-bundles-25660"},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.ver, func(t *testing.T) {
+			actual := getUpstreamBundlesVerDir(tc.ver)
+			if actual != tc.exp {
+				t.Errorf("expected %s on input %s but got %s", tc.exp, tc.ver, actual)
+			}
+		})
+	}
+}
+
+func TestGetUpstreamBundlesPath(t *testing.T) {
```

Comment:
Hmm I like that this increases coverage, but this is actually just testing if `LoadDefaultsForPath()` and `filepath.Join()` works, & the only test case that would actually succeed in a real run is the 25660 one. The rest will fail to download bundles (granted later on in a real mixer run), so I think it's a little over-engineered :) if it writes a good version w/ a good prefix, I think we can safely say it works.

Reply:
Yes, right now it only tests `filepath.Join` because `getUpstreamBundlesPath` is just a wrapper for that function. This is a unit test - part of the value of a unit test is to future-proof any changes we make to that function. We are now testing what we expect from that function and if we change that function that causes it to deviate from our expectations we will be alerted.

Also, what does it cost us to have this unit test? We gain value (validate future changes) and it costs us nothing.